### PR TITLE
Add type hints to SQLite3

### PR DESCRIPTION
**Pull request by Kludex**
Let's try again.

Taking the opportunity to open this PR while working on https://github.com/pydantic/logfire/pull/634.

cc @xrmx @emdneto

### CODE REVIEW

**Review by xrmx — APPROVED** [review]

**Comment by xrmx** [comment]
@Kludex CI is red, you have to use Tuple from typing I guess

File: instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py [hunk 1]
```diff
@@ -39,28 +39,35 @@
...
 from opentelemetry.instrumentation.sqlite3.version import __version__
+from opentelemetry.trace import TracerProvider
 
 # No useful attributes of sqlite3 connection object
 _CONNECTION_ATTRIBUTES = {}
 
 _DATABASE_SYSTEM = "sqlite"
 
+SQLite3Connection = TypeVar(
+    "SQLite3Connection", bound=sqlite3.Connection | None
```
Comment (Kludex) [inline instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py:60]: The problem is here.

File: instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py [hunk 2]
```diff
@@ -39,28 +39,35 @@
 ---
 """
 
+from __future__ import annotations
+
 import sqlite3
 from sqlite3 import dbapi2
-from typing import Collection
+from typing import Any, Collection, TypeVar
```
Comment (Kludex) [inline instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py:46]:
```suggestion
from typing import Any, Collection, TypeVar, Union
```

File: instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py (hunk 1 above)
Comment (Kludex) [reply instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py:60]:
```suggestion
    "SQLite3Connection", bound=Union[sqlite3.Connection, None]
```

**Comment by Kludex** [comment]
> @Kludex CI is red, you have to use Tuple from typing I guess

When you use `from __future__ import annotations`, you can do whatever you want on the types part because they are not evaluated at runtime.

The problem was the `|` (pipe) on the `bound=`. That's used at runtime, and it's only available after 3.10 (or 3.9, I don't remember).

**Comment by xrmx** [comment]
> > @Kludex CI is red, you have to use Tuple from typing I guess
> 
> When you use `from __future__ import annotations`, you can do whatever you want on the types part because they are not evaluated at runtime.
> 
> The problem was the `|` (pipe) on the `bound=`. That's used at runtime, and it's only available after 3.10 (or 3.9, I don't remember).

Not sure about that:

```
==================================== ERRORS ====================================
_ ERROR collecting instrumentation/opentelemetry-instrumentation-sqlite3/tests/test_sqlite3.py _
instrumentation/opentelemetry-instrumentation-sqlite3/tests/test_sqlite3.py:19: in <module>
    from opentelemetry.instrumentation.sqlite3 import SQLite3Instrumentor
instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py:50: in <module>
    from opentelemetry.instrumentation.sqlite3.package import _instruments
instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/package.py:16: in <module>
    _instruments: tuple[str, ...] = tuple()
E   TypeError: 'type' object is not subscriptable
```

**Comment by Kludex** [comment]
> > > @Kludex CI is red, you have to use Tuple from typing I guess
> > 
> > 
> > When you use `from __future__ import annotations`, you can do whatever you want on the types part because they are not evaluated at runtime.
> > The problem was the `|` (pipe) on the `bound=`. That's used at runtime, and it's only available after 3.10 (or 3.9, I don't remember).
> 
> Not sure about that:
> 
> ```
> ==================================== ERRORS ====================================
> _ ERROR collecting instrumentation/opentelemetry-instrumentation-sqlite3/tests/test_sqlite3.py _
> instrumentation/opentelemetry-instrumentation-sqlite3/tests/test_sqlite3.py:19: in <module>
>     from opentelemetry.instrumentation.sqlite3 import SQLite3Instrumentor
> instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/__init__.py:50: in <module>
>     from opentelemetry.instrumentation.sqlite3.package import _instruments
> instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/package.py:16: in <module>
>     _instruments: tuple[str, ...] = tuple()
> E   TypeError: 'type' object is not subscriptable
> ```

I'm missing the future annotations on that file...

File: instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/package.py [hunk 3]
```diff
@@ -13,4 +13,4 @@
 # limitations under the License.
```
Comment (Kludex) [inline instrumentation/opentelemetry-instrumentation-sqlite3/src/opentelemetry/instrumentation/sqlite3/package.py:14]:
```suggestion
from __future__ import annotations
```

**Review by aabmass — APPROVED** [review]
Thanks!

**Pull request merged**